### DISCUSSION TITLE
How do I make a pointer to CoilData[n]?

### DISCUSSION THREAD

**Discussion by Andrew-Garage** [eModbus/eModbus#284] [Q&A] [answered]
Hi everyone!
I'm using Arduino IDE 2.0 and eModBus version [Bunch of flowers]. And i'm stuck on simple silly problem.

Here, a lot of CoilData and float values:
```
CoilData H_Coils("      1        0        0        0        0        1        1        0        0        0 \
/* 0_Прожиг  0* | Br_auto|      --|      --|      --|      --|      --|      --|      --|      --|      --|*/ \
                        0        0        0        0        0        0        0        0        0        0 \
/* 1_Дожиг   1* |      --|      --|      --|      --|      --|      --|      --|      --|      --|      --|*/ \
                        0        0        0        0        0        0        0        0        0        0 \
/* 2_Ак-ка   2* | Ac_auto| Ac_Mode|      --|      --|      --|      --|      --|      --|      --|      --|*/ \
                        0        0        0        0        0        0        0        0        0        0 \
/* 3_ВВИ     3* |      AC|HVT_auto|      --|      --|      --|      --|      --|      --|      --|   START|*/ \
                        0        0        0        0        0        0        0        0        0        0 \
/* 4_Сервис  4* |      --|      --|      --|      --|      --|      --|      --|      --|      --|      --|*/ \
                        0        0        0        0        0        0        0        0        0        0 \
/* 5_Конфиг  5* |      --|      --|      --|      --|      --|      --|      --|      --|      --|      --|*/ ");

/**/

//Адрес в Tesla  |       0|       2|       4|       6|       8|      10|      12|      14|      16|      18|    СМЕЩЕНИЕ +2 байта для FLOAT!

float H_Regs[] = {       2,       5,       0,       0,       0,       0,       0,       0,       0,       0, \
/* 0_Прожиг      |Br_Stage| minutes|      --|      --|      --|      --|      --|      --|      --|      --|*/
                         5,       0,       0,       0,       0,       0,       0,       0,       0,       0, \
/* 1_Дожиг    2* | minutes| seconds|      --|      --|      --|      --|      --|      --|      --|      --|*/
                         1,       5,       0,       0,       0,       0,       0,       0,       0,       0, \
/* 2_Ак-ка    4* |Ac_Stage| minutes|      --|      --|      --|      --|      --|      --|      --|      --|*/
                         5,       0,       0,       0,       0,       0,       0,       0,       0,       0, \
/* 3_ВВИ      6* | minutes|      --|      --|      --|      --|      --|      --|      --|      --|      --|*/
                      0x04,    0x04,    0x04,    0x04,    0x04,    0x04,    0x04,       0,       0,       0, \
/* 4_Сервис   8* | PDA_mdl| Burn_st|  ABr_st|   Ac_st| HVT_mdl|  HVT_st|IFACEmdl|      --|      --|      --|*/};
```

Here, i'm trying to take a set of values:
```
typedef struct {
  CoilData *ACDC;
  CoilData *HVT_auto;
  float *minutes;
} Start_parameters_t;

Start_parameters_t HVT_Start_parameters = {.ACDC = &H_Coils + 30, .HVT_auto = &H_Coils + 31, .minutes = &H_Regs[30]}; 
```
Here, i'm making data output and send it to queue, when i press a certain button:
```
  Serial.print("FROM HVT_Start_parameters! ACDC = ");
  Serial.print((bool)&HVT_Start_parameters.ACDC);
  Serial.print(" HVT_auto = ");
  Serial.print((bool)&HVT_Start_parameters.HVT_auto);
  Serial.print(" minutes = ");
  Serial.println(*HVT_Start_parameters.minutes);

  xQueueSend(CAN_rx_queue, &HVT_Start_parameters, 1000);
```
Here, a code, where i'm getting queue-message. As you i'm also print values directly from DataCoils and float arr.
```
    xQueueReceive(CAN_rx_queue, &HVT_Start_parameters2, portMAX_DELAY);
    Serial.print("OD!  ACDC = ");
    Serial.print(H_Coils[30]);
    Serial.print("  ");
    Serial.print("HVT_auto = ");
    Serial.print(H_Coils[31]);
    Serial.print("  ");
    Serial.print("minutes = ");
    Serial.println(H_Regs[30]);

    Serial.print("HVT_Start_parameters2! ACDC = ");
    Serial.print((bool)&HVT_Start_parameters2.ACDC);
    Serial.print(" HVT_auto = ");
    Serial.print((bool)&HVT_Start_parameters2.HVT_auto);
    Serial.print(" minutes = ");
    Serial.println(*HVT_Start_parameters2.minutes);
```
Here, what i see in Serial monitor:

```
FROM HVT_Start_parameters! ACDC = 1 HVT_auto = 1 minutes = 5.00
OD!  ACDC = 0  HVT_auto = 0  minutes = 5.00
HVT_Start_parameters2! ACDC = 1 HVT_auto = 1 minutes = 5.00
```

ACDC and HVT_auto should be 0. But it's not. Please help me to find what i missed. I've tried a lot of constructions to make correct ref on CoilData. But compilier gives errors.

Full code, if it helps:
https://github.com/Andrew-Garage/Laboratory/blob/3732151017fdf9bfdcddeba5910e7ef10d20fe24/CANOpen_IFACE.ino

**Comment by Miq1** [comment, marked as answer]
``CoilData`` is no elementary data type, the storage locations are handled internally. Therefore pointer arithmetics like you are doing here will not work.
What you can do is hold a pointer to the ``CoilData`` object plus an index number to the coil bit you are referring. This coil then can be read with the ``[]`` operator.

**Reply by Andrew-Garage** [reply]
Could you write an example? I understand that CoilData is not an elementary data type, but due to the fact that I'm a beginner, I don't really understand what you mean. By the way, this code
`Start_parameters_t HVT_Start_parameters = {.ACDC = &H_Coils + 30, .HVT_auto = &H_Coils + 31, .minutes = &H_Regs[30]};`
was suggested by ChatGPT :)

**Reply by Miq1** [reply]
OMG 🙄  Better quickly forget about ChatGPT as a true programmer's helper. It only can suggest what it has seen before and just on a probability base. If it did not have the special application in its training data - which is unlikely in case of eModbus - it will make up something.
The only way is to really learn yourself...

The code you have will in fact calculate a pointer to a ``CoilData`` object that is located 30 times the size of a ``CoilData`` behind the one you have defined -  and that location is rather deep in the woods. You are lucky in that it is still pointing to some existent storage; else your program would simply crash.

The only way to safely address a single coil is by something like ``H_Coils[30]``. So instead of using a pointer at all, why don't you use ``const uint16_t ACDC(30);`` and further down ``Serial.print(H_Coils[ACDC]);``?

**Reply by Andrew-Garage** [reply]
))) I agree with you that chatGPT is not a real assistant. it's just that sometimes it's faster to remember some basic things of the C language with it than to Google it again. it's funny when he doesn't know the answer, he can refer to the data that he invented himself to refer to them during the answer. at least it seems so.
I wanted to make an elegant solution without additional strings, so I wanted to solve with pointers. and I REALLY don't want to redo the CoilData implementation :)) so if you think this is the only right way, then fine.
BTW, ChatGPT quickly explained to me why `const uint16_t ACDC(30);` is much better then  `#define ACDC 30`

**Reply by Miq1** [reply]
😄
